### DISCUSSION TITLE
How to merge PRs created by developers

### DISCUSSION THREAD

**Discussion by josecelano** [Nautilus-Cyberneering/git-queue#207] [General]
From: https://github.com/Nautilus-Cyberneering/git-queue/issues/38#issuecomment-1123975232

We were discussing the process of merging standard PRs for two reasons:

1. We would like to keep the original contributor commit signature.
2. In the future we would like to have at least two branches (GitFlow: `main` and `develop`) and use only bot signatures when merging into `main` or other automatically managed branches.

Thre draft process could be:

- Contributor forks the repo.
- Contributor creates a new branch issue-xxxx-description in the fork.
- Contributor creates commits and push to the branch in the fork.
- Contributor creates a PR into main in the upstream repo.
- If it's all ok we merge with:

```s
git checkout main
git pull
git checkout -b contributor-fork-issue-xxxx-description main
git pull https://github.com/contributor/git-queue.git issue-xxxx-description
git checkout main
git merge --no--ff contributor-fork-issue-xxxx-description
git push origin main
```

If there are new commis in the main branch:

Contributor has to rebase their branch (signing new commits with their GPGP keys).
Go back to the previous state and merge.
The commit is signed with the Nautilus bot GPG key and shown as partially verified.

~~With git `merge --ff-only` we would not lose the contributor's commit signature if we did not use the bot signature. Right now we get partially verified commits because the commit changes because we sign it again with a different GPG key (and we change the committer).~~

UPDATE:

With git `merge` we do not lose the contributor's commit signature.

**Comment by da2ce7** [comment]
@josecelano

Hello José,

With git-merge you never loose the contributor's commit signatures. It doesn't depend on the merge being a fast-forward or not. With git-rebase, git-cherry-pick, and git-amend where you re-write the commits, and therefore loosing all the related signatures.

When merging a feature into the upstream branch, we should not use the `--ff-only` option, rather we should use the `--no-ff` option: We should always create a merge commit when merging into the main branch. (The user who makes the merge should create the merge commit signed by their key).

When merging the upstream into a downstream branch, we should use the `--ff-only` option.  If this fails (as it should if there is any unmerged commits in the feature branch), then we can use git-rebase to reorder the comments on-top of the upstream work.

**Reply by josecelano** [reply]
Hi @da2ce7 yes sorry, I mixed things up a little. I copy/paste that description from [this comment](https://github.com/Nautilus-Cyberneering/git-queue/issues/38#issuecomment-1123583168).

But then I explained better in the [next comment](https://github.com/Nautilus-Cyberneering/git-queue/issues/38#issuecomment-1123975232).

In my previous comments I was mixing two different things:

1. Standard contributor PR merge.
2. Merge a PR from a bot with queue jobs and commits related to jobs.

Some clarifications:

- In the first case we want to keep the contributor's signature. So as you said we would like quite the opposite "--no-ff".
- We are not using PR to merge commits from workers, but if we did it we would need to merge the PR branch using "--ff-only" because it's the only way to detect job conflicts whether you use merge or rebase.

This was my clarification on the next [comment](https://github.com/Nautilus-Cyberneering/git-queue/issues/38#issuecomment-1123975232).

![image](https://user-images.githubusercontent.com/58816/168576246-ab4748f0-e6fb-49a7-8d15-5e8d5a2833b8.png)
